### ISSUE TITLE
Apps lagging when opened in external monitor set as primary display

### ISSUE THREAD

**Issue by jofftiquez** [pop-os/pop#2417]
**Distribution (run `cat /etc/os-release`):**

> NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os

**Related Application and/or Package Version (run `apt policy $PACKAGE NAME`):**

> brave-browser:
  Installed: 1.37.116
  Candidate: 1.37.116
  Version table:
 *** 1.37.116 100
        100 /var/lib/dpkg/status

**Issue/Bug Description:**

I am using a laptop with a 2nd monitor. After the latest update, whenever I set the 2nd/external monitor as the Primary Display, almost all of the apps, most notably brave, discord, etc are lagging. But when I drag them to the laptop's monitor/display the problem is gone. 

Note that this only happened after the latest version

To sum up:

![image](https://user-images.githubusercontent.com/8638243/168501141-ba0f69ae-8b34-450e-a52e-9f697fc0b367.png)

**Steps to reproduce (if you know):**

1. Make external monitor the primary monitor
2. Open brave or other chrome apps in external monitor

**Expected behavior:**

The apps should not lag on either monitors

**Other Notes:**

This has been confirmed in 2 different monitors.

**Comment by jofftiquez** [comment]
Update:

It seems that the only apps lagging are chromium apps, Brave, VSCode, Vivaldi. Not happening in Firefox, terminal, file explorer, etc.

**Comment by nuuttitoivola** [comment]
**Distribution (run `cat /etc/os-release`):**
> NAME="Pop!_OS"
> VERSION="21.10"
> ID=pop
> ID_LIKE="ubuntu debian"
> PRETTY_NAME="Pop!_OS 21.10"
> VERSION_ID="21.10"
> HOME_URL="https://pop.system76.com"
> SUPPORT_URL="https://support.system76.com"
> BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
> PRIVACY_POLICY_URL="https://system76.com/privacy"
> VERSION_CODENAME=impish
> UBUNTU_CODENAME=impish
> LOGO=distributor-logo-pop-os

**Related Application and/or Package Version (run `apt policy $PACKAGE NAME`):**
> $ apt policy discord
> discord:
>   Installed: 0.0.17
>   Candidate: 0.0.17
>   Version table:
>  *** 0.0.17 500
>         500 http://apt.pop-os.org/proprietary impish/main amd64 Packages
>         100 /var/lib/dpkg/status

**Also info on Steam Flatpak (`flatpak info com.valvesoftware.Steam`):** 
```

Steam - Launcher for the Steam software distribution service

          ID: com.valvesoftware.Steam
         Ref: app/com.valvesoftware.Steam/x86_64/stable
        Arch: x86_64
      Branch: stable
     Version: 1.0.0.74
     License: LicenseRef-proprietary
      Origin: flathub
  Collection: org.flathub.Stable
Installation: user
   Installed: 41.7 MB
     Runtime: org.freedesktop.Platform/x86_64/21.08
         Sdk: org.freedesktop.Sdk/x86_64/21.08

      Commit: 8bb8984c379ba159da39079071fde6a0f87e2e40b45a46ff8e77e7f0bf1bd99c
      Parent: 4a606a219c4745d5ce93b290e54395be9031bd39f347d7f73257566abf568085
     Subject: wrapper: Run Steam with -no-cef-sandbox (93f2ab02)
        Date: 2022-02-18 19:38:24 +0000
```

**Issue/Bug Description:**

I have the same issue as the original post after upgrading to the latest Pop!_OS 22.04 LTS. When external monitor is set as primary some apps will lag when used on that monitor, but work normal when used on laptop monitor. Changing primary monitor to laptop "fixes" the issue. 

After trying to figure out the cause for this issue I noticed that the same behavior is reproducible also on Pop!_OS 21.10 **after installing updates** when doing a fresh install. So perhaps the issue stems from somewhere else than Pop itself?

To add to the original post I noticed that if using the external monitor as the only display, then all apps and the Pop!_OS UI are laggy, not just some apps (at least with 21.10). This behavior can be achieved either disabling the laptop monitor from settings or simply closing the laptop lid. 

**Steps to reproduce (if you know):**

1. a) Make external monitor the primary monitor.
2. a) Open up Steam or Discord on the external monitor.
----
1. b) Make external monitor as the only monitor.
2. b) Open up or do anything on the external monitor.

**Expected behavior:**

No app should have lag no matter what is the primary display.

**Other Notes:** 

- I tested that Discord has lag issues both when it is installed as a Flatpak and from a Pop repository.
- I use a `HP E223d` monitor via USB-C cable with my `Asus ROG Zephyrus G14`, by default the USB-C cable transfers the following from the monitor: power, display, webcam & its microphone, ethernet, keyboard, mouse. But unplugging everything from the monitor except power and display didn't do any difference to the lag issue.
- I also did a fresh install of Fedora 36 and it had the same issues of external monitor lagging.
- My guess is that this issue is the same or has something to do with these: https://gitlab.freedesktop.org/xorg/xserver/-/issues/1028 & https://bugs.launchpad.net/xorg-server/+bug/1875015 

Thank you for your work and please ask if you need any more info! :)

**Comment by irishgordo** [comment]
I'm noticing this too.

I just happened to run an upgrade, the pop/system76 that were upgraded were:
```
2022-05-16 09:31:42 upgrade ubuntu-drivers-common:amd64 1:0.9.6.1pop0~1651070156~22.04~66ae311 1:0.9.6.1pop0~1652303087~22.04~1aab0f9

2022-05-16 09:31:45 upgrade libwireplumber-0.4-0:amd64 0.4.9-1pop1~1651700425~22.04~cae0c2c 0.4.9-1pop1~1652301474~22.04~87a71fb

2022-05-16 09:31:45 upgrade wireplumber:amd64 0.4.9-1pop1~1651700425~22.04~cae0c2c 0.4.9-1pop1~1652301474~22.04~87a71fb

2022-05-16 09:31:45 upgrade pop-desktop:amd64 1.6.0~1651773560~22.04~0b475db 1.6.0~1652377940~22.04~67ccd47

2022-05-16 09:31:45 upgrade gnome-shell-extension-cosmic-dock:all 72~ubuntu2pop0~1650928261~22.04~8f33b3b 72~ubuntu2pop0~1652213163~22.04~6cd8fff

2022-05-16 09:31:46 upgrade libpop-support-panel:amd64 1.0.0~1651695094~22.04~5f07d20 1.0.0~1652219183~22.04~8aedc1f

2022-05-16 09:31:46 upgrade pop-icon-theme:all 3.0.0~1650300304~22.04~28b598d 3.1.0~1652288720~22.04~85ab35a

2022-05-16 09:31:48 upgrade pop-launcher:amd64 1.2.1~1648660013~22.04~0893990 1.2.1~1652392579~22.04~47852e5

2022-05-16 09:31:48 upgrade pop-shell:all1 .1.0~1651514427~22.04~15e7acf 1.1.0~1652018335~22.04~add07f1

2022-05-16 09:31:48 upgrade system76-power:amd64 1.1.20~1651773526~22.04~62d9ef1 1.1.20~1652303057~22.04~fc1e801
2022-05-16 09:31:58 upgrade system76-firmware-daemon:amd64 1.0.33~1651262900~22.04~3221872 1.0.36~1652101700~22.04~ac1ba2a

2022-05-16 09:31:58 upgrade system76-scheduler:amd64 1.1.0~1651696814~22.04~cbccc2e 1.1.0~1652366781~22.04~deb28e5

2022-05-16 09:31:58 upgrade system76-driver:all 20.04.50~1650301328~22.04~ab5bdf1 20.04.52~1652101817~22.04~54fc7d4

2022-05-16 09:31:59 upgrade system76-driver-nvidia:all 20.04.50~1650301328~22.04~ab5bdf1 20.04.52~1652101817~22.04~54fc7d4
```

My NVIDIA setup is:
```
Mon May 16 10:13:11 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.103.01   Driver Version: 470.103.01   CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA T1200 La...  Off  | 00000000:01:00.0  On |                  N/A |
| N/A   46C    P5    13W /  N/A |    208MiB /  3911MiB |     28%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      2885      G   /usr/lib/xorg/Xorg                205MiB |
|    0   N/A  N/A      3035      G   ...ome-remote-desktop-daemon        1MiB |
+-----------------------------------------------------------------------------+

```
On a ThinkPad P15 Gen2

However I guess my variant of what I'm seeing is a tad different.

I can make the external display the primary monitor as long as the ThinkPad is not "clamshelled" (with the laptop lid closed).
As soon as the external monitor is the primary monitor and the laptop is closed, the lag is incredibly substantial.

I'm also attaching just the last log of /var/log/dpkg.log - but I believe I've caputured all the system76/popOS programs that were upgraded.

Thanks to all for all the hard work maintaining this!  
I really enjoy using PopOS as my daily driver for work, it's been phenomenal on so many levels!

[last_var_dpkg_upgrade_breaking_things.log](https://github.com/pop-os/pop/files/8702075/last_var_dpkg_upgrade_breaking_things.log)

**Comment by leviport** [comment]
Sounds like this could be related to https://github.com/pop-os/nvidia-graphics-drivers/issues/61

Does performance go back to normal on external displays if you switch from Hybrid to NVIDIA graphics mode?

**Comment by irishgordo** [comment]
fwiw, I think the system76-driver seems just fine, I'm not seeing anything immediately jump out at me between these two commits that I saw:
https://github.com/pop-os/system76-driver/compare/ab5bdf18859bb3d5c024dbc95d0987f6665fbbc7...54fc7d4450c11558d8f8f8d52060a79f7347927c

**Comment by irishgordo** [comment]
@leviport THANK YOU!

Your insight helped inspire me to try to re-select "NVIDIA Graphics" from the GUI.
It was "selected" and I thought to myself "oh hey, yeah, I guess I'm already on NVIDIA Graphics".
But I went ahead and said "oh jeez, what the heck lemme try to select it again".
Fired it off, it did a restart and BANG!
Now both dual 4k displays work great and it went back to setting my one external monitor by default at boot as the primary display.
Buttery smooth!
No lag what-so-ever.

Speaking of making that graphics switch is there a good way to tell what mode you're currently on, if for instance, I feel like the GUI 'though it says I'm on NVIDIA Graphics mode, it might not be"?  Maybe a CLI command to keep tack'd onto a .bashrc / .zshrc that I could fire off post upgrade to 'double check' that I'm really on the correct mode.

**Comment by leviport** [comment]
Glad that helped! I'll close this since it's a bug we're already tracking in https://github.com/pop-os/nvidia-graphics-drivers/issues/61

You can use `system76-power graphics` to query the current graphics mode, but I believe it'll usually start to report the new graphics mode after switching but before reboot. It's really just designed for a reboot right after changing modes.

**Comment by jofftiquez** [comment]
> What is the hardware of these machines? I see one is a P15 gen2.

Razer Blade 14 AMD

**Comment by jofftiquez** [comment]
> Sounds like this could be related to [pop-os/nvidia-graphics-drivers#61](https://github.com/pop-os/nvidia-graphics-drivers/issues/61)
> 
> Does performance go back to normal on external displays if you switch from Hybrid to NVIDIA graphics mode?

Thanks, man! Much appreciated.

**Comment by nuuttitoivola** [comment]
The same solution worked for me also, thanks! 

Weird thing was that these were my settings **before** changing anything: 

![Screenshot from 2022-05-17 06-40-01](https://user-images.githubusercontent.com/26626224/168725177-324ef7a8-f7bf-49f3-ac0e-219d68a58b0b.png)

From the terminal: 
```
$ system76-power graphics
nvidia
```

Now after "switching" they're still the same. It would seem that the system initially gives false info or for some reason the Nvidia has to be set in use even if it's already on by default.

**Comment by jofftiquez** [comment]
> The same solution worked for me also, thanks!
> 
> Weird thing was that these were my settings **before** changing anything:
> 
> ![Screenshot from 2022-05-17 06-40-01](https://user-images.githubusercontent.com/26626224/168725177-324ef7a8-f7bf-49f3-ac0e-219d68a58b0b.png)
> 
> From the terminal:
> 
> ```
> $ system76-power graphics
> nvidia
> ```
> 
> Now after "switching" they're still the same. It would seem that the system initially gives false info or for some reason the Nvidia has to be set in use even if it's already on by default.

Same, the state was the same. It looks like you just really have to explicitly set it again. LOL.

**Comment by irishgordo** [comment]
@jofftiquez @nuuttitoivola 
I should of snagged a screen shot of mine! :laughing: 
But yes, mine was most definitely in the same state and I had to set it again explicitly and power-cycle the laptop.

**Comment by dmitrypro77** [comment]
Confirming, the solution above is working for me as well. I had the same issue and spent a lot of time to figure it out. Thanks!

**Comment by vivaan98** [comment]
> **Distribution (run `cat /etc/os-release`):**
> 
> > NAME="Pop!_OS"
> > VERSION="22.04 LTS"
> > ID=pop
> > ID_LIKE="ubuntu debian"
> > PRETTY_NAME="Pop!_OS 22.04 LTS"
> > VERSION_ID="22.04"
> > HOME_URL="https://pop.system76.com"
> > SUPPORT_URL="https://support.system76.com"
> > BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
> > PRIVACY_POLICY_URL="https://system76.com/privacy"
> > VERSION_CODENAME=jammy
> > UBUNTU_CODENAME=jammy
> > LOGO=distributor-logo-pop-os
> 
> **Related Application and/or Package Version (run `apt policy $PACKAGE NAME`):**
> 
> > brave-browser:
> > Installed: 1.37.116
> > Candidate: 1.37.116
> > Version table:
> > *** 1.37.116 100
> > 100 /var/lib/dpkg/status
> 
> **Issue/Bug Description:**
> 
> I am using a laptop with a 2nd monitor. After the latest update, whenever I set the 2nd/external monitor as the Primary Display, almost all of the apps, most notably brave, discord, etc are lagging. But when I drag them to the laptop's monitor/display the problem is gone.
> 
> Note that this only happened after the latest version
> 
> To sum up:
> 
> ![image](https://user-images.githubusercontent.com/8638243/168501141-ba0f69ae-8b34-450e-a52e-9f697fc0b367.png)
> 
> **Steps to reproduce (if you know):**
> 
> 1. Make external monitor the primary monitor
> 2. Open brave or other chrome apps in external monitor
> 
> **Expected behavior:**
> 
> The apps should not lag on either monitors
> 
> **Other Notes:**
> 
> This has been confirmed in 2 different monitors.

hey, i have the exact same problem but everything is lagging for me

**Comment by Leandro-Michail-Krikis** [comment]
Another possible solution for chromium based apps is to right click and select "launch using discrete graphics card"

Switching to nvidia graphics wasn't possible for me, I need the battery life for the university. In hybrid graphics battery life it is even better than windows.

Ryzen 5800h
RTX 3050
In my laptop the HDMI port is connected directly in the GPU, I think this is the culprit of this bug

**Comment by rahulp3** [comment]
> @leviport THANK YOU!
> 
> Your insight helped inspire me to try to re-select "NVIDIA Graphics" from the GUI. It was "selected" and I thought to myself "oh hey, yeah, I guess I'm already on NVIDIA Graphics". But I went ahead and said "oh jeez, what the heck lemme try to select it again". Fired it off, it did a restart and BANG! Now both dual 4k displays work great and it went back to setting my one external monitor by default at boot as the primary display. Buttery smooth! No lag what-so-ever.
> 
> Speaking of making that graphics switch is there a good way to tell what mode you're currently on, if for instance, I feel like the GUI 'though it says I'm on NVIDIA Graphics mode, it might not be"? Maybe a CLI command to keep tack'd onto a .bashrc / .zshrc that I could fire off post upgrade to 'double check' that I'm really on the correct mode.

Even i had the similar issue and re-selecting the "NVIDIA Graphics"  option worked for me as well.
Thanks!
